Introduce support for environments where the standard library is not available (#![no_std]) but a global allocator is

IntegralPilot opened the pull request:
This PR introduces support for environments where the standard library is not available but a global allocator is (i.e. embedded usecase) to the "subcrates" in this repository - `wain-ast`, `wain-syntax-text`, `wain-syntax-binary`, `wain-validate` and `wain-exec`.

**`wain-ast`**:

- Support added without needing a feature to be enabled.
- Incremented semver "minor" version so package version is now `0.3.0`.

**`wain-exec`**:

- Support added and locked behind `no_std` non-default feature.
- This is because to add support a dependency is required on `libm` - this is only included with the `no_std` feature to make the claim that "no other dependencies are used" true for the default featureset.
- Incremented semver "minor" version so package version is now `0.4.0`.
- `execute` function and `DefaultImporter` are not available when the `no_std` feature is enabled - developers who wish to use this feature need to create their own `Importer` and use runtime APIs to create and run a `Runtime` that uses it.

**`wain-syntax-binary`**:
- Support added without needing a feature to be enabled.
- Incremented semver "minor" version so package version is now `0.2.0`.

**`wain-syntax-text`**:
- Support added and locked behind `no_std` non-default feature.
- This is because to add support a dependency is required on `libm` and `hashbrown` - this is only included with the `no_std` feature to make the claim that "no other dependencies are used" true for the default featureset.
- Incremented semver "minor" version so package version is now `0.3.0`.

**`wain-validate`**:
- Support added and locked behind `no_std` non-default feature.
- This is because to add support a dependency is required on `hashbrown` - this is only included with the `no_std` feature to make the claim that "no other dependencies are used" true for the default featureset.
- Incremented semver "minor" version so package version is now `0.2.0`.

Additionally, post-commit hooks were failing due to Clippy identified "dead code" in `spec-test`. I did not alter the code in `spec-test` to introduce `#![no_std]` support however have commented out the code that turned out to be actually dead and added Clippy ignore directives to the code that was being used in the tests but not detected by Clippy, in order to get these checks to pass and resolve all Clippy warnings.

If you decide to approve this PR, can you please deploy the new versions of the crates to `crates.io` so that I can use them more easily in my project?

Thank you for your time!

IntegralPilot committed:
make wain-ast #![no_std] compatible

IntegralPilot committed:
add support for #![no_std] in wain-syntax-binary

IntegralPilot committed:
#![no_std] for wain-syntax-text

IntegralPilot committed:
#![no_std] support for wain-validate

IntegralPilot committed:
make it so the claim of 'no other dependencies' is true when the no_std features are not enabled

IntegralPilot committed:
add #![no_std] support to wain-exec

IntegralPilot committed:
increment MINOR semver field of each subcrate to indicate new functionality (no_std support) without breaking existing API

rhysd reviewed (changes requested):
Thank you. Supporting `no_std` would be useful. Before reviewing details in this branch, let me request some points:

- Would you apply the following patch and revert the code commented out due to warnings?
  ```diff
  diff --git a/spec-test/src/main.rs b/spec-test/src/main.rs
  index ddbd226..d502206 100644
  --- a/spec-test/src/main.rs
  +++ b/spec-test/src/main.rs
  @@ -1,4 +1,5 @@
   #![forbid(unsafe_code)]
  +#![allow(dead_code)]

   mod crash;
   mod error;

  ```
- Would you revert the package version bumps in this PR? I will do them in my side after understanding the changes. Since the major versions are zero, patch versions should be bumped if there are no breaking changes.
- Could you run [the spec test](https://github.com/rhysd/wain/blob/master/spec-test/README.md) on this branch with enabling `no_std` and compare the results with `main` branch? If the numbers of passed test cases don't match, it would mean that there are some mistakes in the implementation added in this branch.
  - On my local machine, the results of the speci tests are:
    ```
    total: 19757, passed: 19460, failed: 179, skipped: 118
    ```

IntegralPilot committed:
revert commenting out of 'dead' code - instead disable the clippy lint

IntegralPilot committed:
revert package version changes

IntegralPilot committed:
no_std testing support

IntegralPilot commented:
Thank you for your reply!

I've reverted the version changes and commenting-out of "dead" code, as well as added `#![allow(dead_code)]` to the `spec-tests`.

With regards to the failing tests numbers, I've added a `no_std` test mode to the spec test (run with `cargo run --features=no_std`) and when running with this mode the test output and numbers of passed/failing/skipped tests are identical.

```
PS C:\Users\<username>\wain\spec-test> diff nostdtests.txt stdtests.txt

InputObject    SideIndicator
-----------    -------------
stdtests.txt   =>
nostdtests.txt <=
```

Is there anything else you would like?

Thanks again!

rhysd commented:
Wow, passing the same spec-tests is great. I'll review the details. Since this PR is a bit larger, let me take a look at it on the weekend.

rhysd commented on wain-exec/src/import.rs:
```diff
@@ -36,17 +40,25 @@ pub fn check_func_signature(
-pub struct DefaultImporter<R: Read, W: Write> {
+/// With `no_std` feature enabled, there is no `DefaultImporter`, it is up to you to create one suitable for your system.
+#[cfg(not(feature = "no_std"))]
+pub struct DefaultImporter<R, W>
```
Question: Why `DefaultImporter` cannot be implemented when `no_std`?

rhysd commented on wain-syntax-text/src/ast.rs:
```diff
@@ -1,6 +1,10 @@
+use alloc::fmt;
+use alloc::{borrow::Cow, vec::Vec};
+#[cfg(feature = "hashbrown")]
+use hashbrown::HashMap;
```
`alloc::collections::BTreeMap` is available. Can we use `BTreeMap` instead of introducing a new dependency `hashbrown` on `no_std`? Or should we still use hashbrown?

https://github.com/rust-lang/rust/blob/80d8270d8488957f62fbf0df7a19dfe596be92ac/library/alloc/src/collections/mod.rs#L39

rhysd commented:
Closing as no response.
